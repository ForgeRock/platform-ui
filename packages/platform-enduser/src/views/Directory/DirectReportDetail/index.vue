<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BContainer class="mt-5">
    <BMedia
      no-body
      data-testid="detail-report-header">
      <BMediaAside vertical-align="start">
        <BImg
          class="rounded-circle"
          height="104"
          width="104"
          :alt="directReportUserInfo.name"
          :aria-hidden="true"
          :src="directReportUserInfo.profileImage || require('@forgerock/platform-shared/src/assets/images/avatar.png')" />
      </BMediaAside>
      <BMediaBody class="text-truncate">
        <FrHeader
          :top-text="$t('common.user.user')"
          :title="directReportUserInfo.name"
          :subtitle="directReportUserInfo.userName" />
      </BMediaBody>
    </BMedia>
    <BCard
      no-body
      class="card-tabs-vertical">
      <BTabs
        v-if="tabIndex !== null"
        v-model="tabIndex"
        data-testid="access-tabs"
        @activate-tab="tabActivated"
        nav-wrapper-class="d-none d-md-block"
        content-class="overflow-hidden position-inherit"
        pills
        vertical>
        <BTab
          v-for="(tab, index) in tabItems"
          class="nav-item"
          :active="tab.active"
          :key="`${tab.grantType}_${index}`"
          :title="tab.displayName"
          :title-link-attributes="{'data-testid':`access-tab-${tab.grantType}`}">
          <template #title>
            <FrIcon
              class="mr-3"
              :name="tab.icon" />
            {{ tab.displayName }}
          </template>
          <FrMyAccessReviewTable
            v-if="directReportUserInfo.userId"
            :default-sort="getDefaultSort(tab.grantType)"
            :fields="getTableFields(tab.grantType)"
            :grant-type="tab.grantType"
            resource-name="directReportDetail"
            :user-id="directReportUserInfo.userId" />
        </BTab>
      </BTabs>
    </BCard>
  </BContainer>
</template>

<script>
import {
  BCard,
  BContainer,
  BImg,
  BMedia,
  BMediaAside,
  BMediaBody,
  BTabs,
  BTab,
} from 'bootstrap-vue';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import BreadcrumbMixin from '@forgerock/platform-shared/src/mixins/BreadcrumbMixin';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import FrMyAccessReviewTable from '../../MyAccessReview/MyAccessReviewTable';
import { getDirectReportUserInfo } from '@/api/GovernanceEnduserApi';

/**
 * DirectReports base container component. This component renders a vertical tab menu for application, entitlement and role options
 * Each tab displays corresponding list of My Access attributes defined in the system.
 */
export default {
  name: 'DirectReportDetail',
  components: {
    BCard,
    BContainer,
    BImg,
    BMedia,
    BMediaAside,
    BMediaBody,
    BTabs,
    BTab,
    FrHeader,
    FrIcon,
    FrMyAccessReviewTable,
  },
  mixins: [
    BreadcrumbMixin,
    NotificationMixin,
  ],
  async created() {
    await this.getUserProfile();
  },
  data() {
    return {
      tabItems: [
        {
          displayName: this.$t('common.accounts'),
          icon: 'apps',
          routeName: 'Application',
          grantType: 'account',
        },
        {
          displayName: this.$t('pages.myAccess.entitlement.tabTitle'),
          icon: 'assignment_turned_in',
          routeName: 'Entitlement',
          grantType: 'entitlement',
        },
        {
          displayName: this.$t('pages.myAccess.role.tabTitle'),
          icon: 'assignment_ind',
          routeName: 'Role',
          grantType: 'role',
        },
      ],
      tabIndex: null,
      directReportUserInfo: {},
    };
  },
  mounted() {
    this.setBreadcrumb('/my-reports', this.$t('governance.directReports.title'));
    this.tabItems.find((item, index) => {
      if (item.grantType === this.$route.params.grantType) {
        item.active = true;
        this.tabIndex = index;
        return true;
      }
      return false;
    });
  },
  methods: {
    getDefaultSort(grantType) {
      switch (grantType) {
        case 'account':
        case 'entitlement':
        default:
          return 'application.name';
        case 'role':
          return 'role.name';
      }
    },
    getTableFields(grantType) {
      switch (grantType) {
        case 'account':
          return [{
            key: 'appName',
            label: this.$t('common.name'),
            sortable: true,
          },
          {
            key: 'accountName',
            label: this.$t('pages.myAccess.accountName'),
            sortable: true,
          }];
        case 'role':
          return [
            {
              key: 'roleName',
              label: this.$t('common.name'),
              sortable: true,
            },
            {
              key: 'timeConstraint',
              label: this.$t('pages.myAccess.timeConstraint'),
            },
          ];
        case 'entitlement':
          return [
            {
              key: 'appName',
              label: this.$t('common.application'),
              sortable: true,
            },
            {
              key: 'entitlementName',
              label: this.$t('common.name'),
              sortable: true,
            },
            {
              key: 'accountName',
              label: this.$t('pages.myAccess.accountName'),
              sortable: true,
            },
          ];
        default:
          return [];
      }
    },
    async getUserProfile() {
      await getDirectReportUserInfo(this.$store.state.UserStore.userId, this.$route.params.userId).then(({ data }) => {
        this.directReportUserInfo = {
          name: `${data.givenName} ${data.sn}`,
          userName: data.userName,
          userId: data._refResourceId,
          profileImage: data.profileImage,
        };
      }).catch((err) => {
        this.showErrorMessage(err, this.$t('governance.directReports.errorGettingDirectReportUserInfo'));
      });
    },
    tabActivated(tabIndex) {
      if (tabIndex > -1) {
        this.$route.params.grantType = this.tabItems[tabIndex].grantType;
        const { grantType, userId } = this.$route.params;
        window.history.pushState(window.history.state, '', `#/my-reports/${userId}/${grantType}`);
      }
    },
  },
};
</script>
