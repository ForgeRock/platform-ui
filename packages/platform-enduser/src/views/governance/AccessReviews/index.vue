<!-- Copyright (c) 2023-2025 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BContainer>
    <template v-if="tableLoading">
      <FrSpinner class="py-5" />
    </template>
    <div
      v-else
      class="my-5">
      <FrHeader
        :title="$t('pages.accessReview.title')"
        :subtitle="$t('pages.accessReview.subTitle')" />
      <BCard no-body>
        <BCardHeader class="p-0">
          <div class="btn-toolbar justify-content-between p-3 border-bottom-0">
            <div>
              <BDropdown
                id="dropdown-status"
                variant="link"
                class="ml-2 text-decoration-none">
                <template #button-content>
                  <span
                    class="font-weight-bold mr-1"
                  >
                    {{ `${$t('common.status')}:` }}
                  </span>
                  <span>{{ statusSort.text }}</span>
                </template>
                <template v-for="([key, statusObj]) in Array.from(statuses)">
                  <BDropdownItem
                    :key="key"
                    v-if="!statusObj.sortHide && !statusObj.sortAccessReviewHide"
                    @click="handleStatusSort(statusObj)"
                  >
                    <span
                      :class="[{'pl-3': (statusObj.type !== statuses.get('closed').type && statusObj.type !== statuses.get('in-progress').type)}]"
                    >
                      {{ statusObj.text }}
                    </span>
                  </BDropdownItem>
                </template>
              </BDropdown>
            </div>
          </div>
        </BCardHeader>
        <BTable
          v-if="accessReviewList.length"
          class="mb-0"
          hover
          responsive
          tbody-tr-class="cursor-pointer"
          primary-key="id"
          :fields="fields"
          :items="accessReviewList"
          @row-clicked="viewApplicationDetails">
          <template #cell(name)="{ item }">
            <BMedia
              class="align-items-center"
              no-body>
              <div class="media-body align-self-center">
                <h5 class="m-0">
                  {{ item.campaignName || item.name }}
                </h5>
                <small class="text-muted">
                  {{ item.description }}
                </small>
              </div>
            </BMedia>
          </template>
          <template #cell(status)="{ item }">
            <template v-if="item && item.status">
              <BBadge
                class="w-100"
                :variant="statuses.get(item.status).variant">
                {{ $t(`pages.accessReview.status.${item.status.toLowerCase()}`) }}
              </BBadge>
            </template>
          </template>
          <template #cell(progress)="{ item }">
            <FrCircleProgressBar
              :id="`taskProgress${item.id}`"
              :progress="progressPercentage(item.progress)"
              :thickness="6"
              :empty-thickness="6"
              :empty-color="styles.brightGray"
              :size="50">
              <template #count="{ count }">
                <small class="font-weight-bold mb-0">
                  {{ count.currentValue }}%
                </small>
              </template>
            </FrCircleProgressBar>
            <BPopover
              triggers="hover"
              boundary="window"
              placement="top"
              :target="`taskProgress${item.id}`">
              <div class="p-1">
                <h5 class="mb-1">
                  {{ $t('pages.accessReview.taskPopoverCompletedText', { percentage: progressPercentage(item.progress) }) }}
                </h5>
                {{ $t('pages.accessReview.taskPopoverItemsCompletedText', { completed: item.totals.completed, total: item.totals.total }) }}
              </div>
            </BPopover>
          </template>
        </BTable>
        <FrNoData
          v-else
          :card="false"
          class="mb-4"
          data-testid="access-review-no-data"
          icon="inbox"
          :subtitle="$t('governance.certificationTask.noAccessReview', { type: statusSort.text })" />
        <FrPagination
          v-if="totalRows"
          :value="currentPage"
          :total-rows="totalRows"
          :per-page="pageSize"
          @input="paginationChange"
          @on-page-size-change="pageSizeChange" />
      </BCard>
    </div>
  </BContainer>
</template>

<script>
import { mapState } from 'pinia';
import { useUserStore } from '@forgerock/platform-shared/src/stores/user';
import {
  BBadge,
  BCard,
  BCardHeader,
  BContainer,
  BDropdown,
  BDropdownItem,
  BMedia,
  BPopover,
  BTable,
} from 'bootstrap-vue';
import {
  getCertificationItems,
  searchCertificates,
  forwardCertification,
} from '@forgerock/platform-shared/src/api/governance/CertificationApi';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import FrCircleProgressBar from '@forgerock/platform-shared/src/components/CircleProgressBar';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
import FrNoData from '@forgerock/platform-shared/src/components/NoData';
import FrPagination from '@forgerock/platform-shared/src/components/Pagination';
import FrSpinner from '@forgerock/platform-shared/src/components/Spinner';
import CertificationMixin from '@forgerock/platform-shared/src/mixins/Governance/Certification';
import styles from '@/scss/main.scss';

export default {
  name: 'AccessReviews',
  components: {
    BBadge,
    BCard,
    BCardHeader,
    BContainer,
    BDropdown,
    BDropdownItem,
    BMedia,
    BPopover,
    BTable,
    FrCircleProgressBar,
    FrHeader,
    FrNoData,
    FrPagination,
    FrSpinner,
  },
  data() {
    return {
      pageSize: 10,
      currentPage: 1,
      totalRows: 0,
      tableLoading: true,
      selectedCertId: null,
      accessReviewList: [],
      fields: [
        {
          key: 'name',
          label: this.$t('common.name'),
        },
        {
          key: 'formattedDeadline',
          label: this.$t('common.deadline'),
          sortable: true,
        },
        {
          key: 'status',
          class: 'w-140px',
          label: this.$t('common.status'),
        },
        {
          key: 'progress',
          class: 'w-140px',
          label: this.$t('common.progress'),
        }],
      styles,
    };
  },
  mixins: [
    CertificationMixin,
    NotificationMixin,
  ],
  mounted() {
    this.getList();
  },
  computed: {
    ...mapState(useUserStore, ['userId']),
    currentUserId() {
      return `managed/user/${this.userId}`;
    },
  },
  methods: {
    pageSizeChange(newPageSize) {
      this.pageSize = newPageSize;
      this.currentPage = 1;
      this.getList();
    },
    paginationChange(pageNumber) {
      this.currentPage = pageNumber;
      this.getList();
    },
    getItems(params) {
      const { queryString } = params;
      delete params.queryString;
      if (!queryString) {
        return getCertificationItems(params);
      }
      return searchCertificates(queryString, params);
    },
    viewApplicationDetails(rowData) {
      if (!rowData || !rowData.campaignId || !rowData.certifierId) {
        this.showErrorMessage('', this.$t('governance.certificationTask.errors.certificationDetailError'));
      } else {
        this.$router.push({
          name: 'CertificationTask',
          params: { campaignId: rowData?.campaignId },
          query: {
            actorId: rowData?.certifierId,
            taskStatus: this.statusSort.param,
          },
        });
      }
    },
    progressPercentage(progress = 0) {
      const actualProgress = progress * 100;
      return actualProgress >= 50
        ? Math.floor(actualProgress)
        : Math.ceil(actualProgress);
    },
    /**
     * @param {Object} data data of the certification to forward
     * @param {String} data.certId id of the certification to forward
     * @param {String} data.newActorId id of new actor
     * @param {String} data.comment comment for the forward
     */
    forward({ certId, newActorId, comment = '' }) {
      forwardCertification(this.currentUserId, certId, newActorId, comment)
        .then(() => {
          this.displayNotification('success', this.$t('governance.certificationTask.success.accessReviewForwarded'));
          this.$root.$emit('bv::hide::modal', 'forward-review-modal');
          this.getList();
        })
        .catch((error) => {
          this.showErrorMessage(error, this.$t('governance.certificationTask.errors.forwardItemError'));
        });
    },
  },
};
</script>
<style lang="scss" scoped>
#dropdown-status:deep {
  button {
    color: $black;
    padding-left: 0;
    padding-right: 0;
  }
}
:deep {
  .w-140px {
    width: 140px;
  }
 .w-96px {
    width: 96px;
  }
}
</style>
