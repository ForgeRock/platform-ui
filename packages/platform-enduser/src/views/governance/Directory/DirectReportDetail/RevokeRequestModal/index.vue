<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BModal
    id="revoke-request-modal"
    size="lg"
    title-class="h5"
    title-tag="h2"
    :static="isTesting"
    :title="$t('governance.directReports.revokeRequest')"
    @hidden="$emit('hidden')">
    <!-- Justification textarea -->
    <BFormGroup>
      <FrField
        v-model="justificationText"
        data-testid="justification-field"
        type="textarea"
        :label="$t('governance.directReports.justification')"
        :description="$t('governance.directReports.revokeJustification')"
        :max-rows="10"
        :rows="5" />
    </BFormGroup>
    <!-- Priority dropdown select -->
    <BFormGroup>
      <div class="mb-2 text-muted">
        {{ $t('common.priority') }}
      </div>
      <FrField
        v-model="selectedPriority"
        data-testid="priority-field"
        name="Priority"
        type="select"
        :options="priorityOptions"
        :preselect-first="true"
        :searchable="false">
        <template
          v-for="(slotName, index) in ['option', 'singleLabel']"
          #[slotName]="{ option }">
          <span
            :key="index"
            class="d-flex align-items-center">
            <BImg
              class="mr-2"
              width="24"
              :alt="option.text"
              :src="option.imgSrc" />
            {{ option.text }}
          </span>
        </template>
      </FrField>
    </BFormGroup>
    <!-- Expiration Datepicker -->
    <BFormGroup>
      <div class="mb-2 text-muted">
        {{ $t('governance.accessRequest.newRequest.applyExpirationDate') }}
      </div>
      <FrDatepicker
        v-model="expirationDate"
        :date-format-options="{ year: 'numeric', month: 'numeric', day: 'numeric' }"
        :placeholder="$t('governance.accessRequest.newRequest.expiryDate')" />
    </BFormGroup>
    <template #modal-footer="{ cancel }">
      <div class="w-100 d-flex justify-content-end">
        <BButton
          variant="link"
          :aria-label="$t('common.cancel')"
          @click="cancel()">
          {{ $t('common.cancel') }}
        </BButton>
        <FrButtonWithSpinner
          data-testid="revoke-request-submit-button"
          variant="primary"
          :button-text="$t('governance.directReports.submitRequest')"
          :spinner-text="$t('common.submitting')"
          :show-spinner="showSpinner"
          @click="submission" />
      </div>
    </template>
  </BModal>
</template>

<script>
import {
  BButton,
  BFormGroup,
  BImg,
  BModal,
} from 'bootstrap-vue';
import FrButtonWithSpinner from '@forgerock/platform-shared/src/components/ButtonWithSpinner/';
import FrDatepicker from '@forgerock/platform-shared/src/components/Datepicker';
import FrField from '@forgerock/platform-shared/src/components/Field';
import getPriorityImageSrc from '@/components/utils/governance/AccessRequestUtils';

/**
 * Provides the ability to revoke grants for direct reports.
 */
export default {
  name: 'RevokeRequestModal',
  components: {
    BButton,
    BFormGroup,
    BImg,
    BModal,
    FrButtonWithSpinner,
    FrDatepicker,
    FrField,
  },
  props: {
    isTesting: {
      type: Boolean,
      default: false,
    },
    showSpinner: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      expirationDate: '',
      justificationText: '',
      priorityOptions: [
        {
          imgSrc: getPriorityImageSrc('low'),
          text: this.$t('governance.accessRequest.newRequest.priority.low'),
          value: 'low',
        },
        {
          imgSrc: getPriorityImageSrc('medium'),
          text: this.$t('governance.accessRequest.newRequest.priority.med'),
          value: 'medium',
        },
        {
          imgSrc: getPriorityImageSrc('high'),
          text: this.$t('governance.accessRequest.newRequest.priority.high'),
          value: 'high',
        },
      ],
      selectedPriority: 'low',
    };
  },
  methods: {
    submission() {
      const payload = { priority: this.selectedPriority };
      if (this.expirationDate) {
        payload.expiryDate = this.expirationDate;
      }
      if (this.justificationText) {
        payload.justification = this.justificationText;
      }
      this.$emit('submission', payload);
    },
  },
};
</script>

<style lang="scss" scoped>
::v-deep .multiselect .multiselect__option--selected:after {
  content: 'check';
  font-family: 'Material Icons Outlined', Serif;
  font-size: 15px;
  color: $green !important;
  padding-top: 0.25rem;
  background: initial;
  padding-left: 10px;
}
</style>
