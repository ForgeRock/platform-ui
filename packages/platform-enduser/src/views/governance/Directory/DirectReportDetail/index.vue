<!-- Copyright (c) 2023-2024 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div>
    <BContainer class="mt-5">
      <BMedia
        no-body
        data-testid="detail-report-header">
        <BMediaAside vertical-align="start">
          <BImg
            class="rounded-circle"
            height="104"
            width="104"
            :alt="directReportUserInfo.name"
            :aria-hidden="true"
            :src="directReportUserInfo.profileImage || require('@forgerock/platform-shared/src/assets/images/avatar.png')" />
        </BMediaAside>
        <BMediaBody class="text-truncate">
          <FrHeader
            :top-text="$t('common.user.user')"
            :title="directReportUserInfo.name"
            :subtitle="directReportUserInfo.userName" />
        </BMediaBody>
      </BMedia>
      <BCard
        no-body
        class="card-tabs-vertical">
        <BDropdown
          v-if="tabItems.length"
          variant="white"
          block
          menu-class="w-100"
          class="d-md-none border-bottom mobile-dropdown-menu">
          <template #button-content>
            <FrIcon
              icon-class="mr-2"
              :name="tabItems[tabIndex].icon">
              {{ pluralizeValue(tabItems[tabIndex].displayName) }}
            </FrIcon>
          </template>
          <BDropdownItem
            v-for="(tab, index) in tabItems"
            :key="index"
            :active="index === tabIndex"
            @click="tabIndex = index">
            <FrIcon
              icon-class="mr-3"
              :name="tab.icon">
              {{ pluralizeValue(tab.displayName) }}
            </FrIcon>
          </BDropdownItem>
        </BDropdown>
        <BTabs
          v-if="tabIndex !== null"
          v-model="tabIndex"
          data-testid="access-tabs"
          @activate-tab="tabActivated"
          nav-wrapper-class="d-none d-md-block"
          content-class="overflow-hidden position-inherit"
          pills
          vertical>
          <BTab
            v-for="(tab, index) in tabItems"
            class="nav-item"
            :active="tab.active"
            :key="`${tab.grantType}_${index}`"
            :title="tab.displayName"
            :title-link-attributes="{'data-testid':`access-tab-${tab.grantType}`}">
            <template #title>
              <FrIcon
                icon-class="mr-3"
                :name="tab.icon">
                {{ tab.displayName }}
              </FrIcon>
            </template>
            <FrGovResourceTable
              v-if="directReportUserInfo.userId && tab.active"
              :fields="getTableFields(tab.grantType)"
              :grant-type="tab.grantType"
              :items="resourceItems"
              :require-request-justification="requireRequestJustification"
              :total-count="resourceTotalCount"
              :modal-id="`${tab.displayName}-modal`"
              @load-data="queryResource"
              @revoke-items="revokeResourcesAndCloseModal($event, directReportUserInfo.userId, `${tab.displayName}-modal-revoke`)" />
          </BTab>
        </BTabs>
      </BCard>
    </BContainer>
  </div>
</template>

<script>
import {
  BCard,
  BContainer,
  BDropdown,
  BDropdownItem,
  BImg,
  BMedia,
  BMediaAside,
  BMediaBody,
  BTabs,
  BTab,
} from 'bootstrap-vue';
import { mapState } from 'pinia';
import { useUserStore } from '@forgerock/platform-shared/src/stores/user';
import { pluralizeValue } from '@forgerock/platform-shared/src/utils/PluralizeUtils';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import useBreadcrumb from '@forgerock/platform-shared/src/composables/breadcrumb';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import { getIgaAccessRequest } from '@forgerock/platform-shared/src/api/governance/CommonsApi';
import FrGovResourceTable from '@forgerock/platform-shared/src/components/governance/GovResourceTable';
import { getGovernanceGrants, revokeResourcesFromIGA } from '@forgerock/platform-shared/src/utils/governance/resource';
import { getDirectReportUserInfo } from '@/api/governance/DirectoryApi';

/**
 * DirectReports base container component. This component renders a vertical tab menu for application, entitlement and role options
 * Each tab displays corresponding list of My Access attributes defined in the system.
 */
export default {
  name: 'DirectReportDetail',
  components: {
    BCard,
    BContainer,
    BDropdown,
    BDropdownItem,
    BImg,
    BMedia,
    BMediaAside,
    BMediaBody,
    BTabs,
    BTab,
    FrHeader,
    FrIcon,
    FrGovResourceTable,
  },
  mixins: [
    NotificationMixin,
  ],
  setup() {
    const { setBreadcrumb } = useBreadcrumb();
    return { setBreadcrumb };
  },
  data() {
    return {
      directReportUserInfo: {},
      requireRequestJustification: false,
      resourceItems: [],
      resourceTotalCount: 0,
      tabItems: [
        {
          displayName: this.$t('common.accounts'),
          icon: 'apps',
          routeName: 'Application',
          grantType: 'account',
        },
        {
          displayName: this.$t('common.entitlements'),
          icon: 'assignment_turned_in',
          routeName: 'Entitlement',
          grantType: 'entitlement',
        },
        {
          displayName: this.$t('pages.myAccess.role.tabTitle'),
          icon: 'assignment_ind',
          routeName: 'Role',
          grantType: 'role',
        },
      ],
      tabIndex: 0,
      userId: this.$route.params.resourceId || useUserStore().userId,
    };
  },
  async created() {
    await this.getUserProfile();
  },
  async mounted() {
    this.setBreadcrumb('/my-reports', this.$t('governance.directReports.title'));
    this.tabItems.find((item, index) => {
      if (item.grantType === this.$route.params.grantType) {
        item.active = true;
        this.tabIndex = index;
        return true;
      }
      return false;
    });
    try {
      const { data } = await getIgaAccessRequest();
      this.requireRequestJustification = data.requireRequestJustification;
    } catch {
      // We don't need to show an error here
    }
  },
  computed: {
    ...mapState(useUserStore, ['userId']),
  },
  methods: {
    pluralizeValue,
    getTableFields(grantType) {
      const sharedFields = [
        {
          key: 'assignment',
          label: this.$t('common.assignment'),
          sortable: false,
          thClass: 'w-150px',
        },
        {
          key: 'actions',
          label: '',
          class: 'p-3',
          sortable: false,
          thClass: 'w-100px',
        },
      ];
      switch (grantType) {
        case 'account':
          return [
            {
              key: 'appName',
              label: this.$t('common.name'),
              sortable: true,
            },
            {
              key: 'accountName',
              label: this.$t('pages.myAccess.accountName'),
              sortable: true,
            },
            ...sharedFields,
          ];
        case 'role':
          return [
            {
              key: 'roleName',
              label: this.$t('common.name'),
              sortable: true,
            },
            {
              key: 'timeConstraint',
              label: this.$t('pages.myAccess.timeConstraint'),
            },
            ...sharedFields,
          ];
        case 'entitlement':
          return [
            {
              key: 'appName',
              label: this.$t('common.application'),
              sortable: true,
            },
            {
              key: 'entitlementName',
              label: this.$t('common.name'),
              sortable: true,
            },
            {
              key: 'accountName',
              label: this.$t('pages.myAccess.accountName'),
              sortable: true,
            },
            ...sharedFields,
          ];
        default:
          return [];
      }
    },
    async getUserProfile() {
      await getDirectReportUserInfo(this.userId, this.$route.params.userId).then(({ data }) => {
        this.directReportUserInfo = {
          name: `${data.givenName} ${data.sn}`,
          userName: data.userName,
          userId: data._refResourceId,
          profileImage: data.profileImage,
        };
      }).catch((err) => {
        this.showErrorMessage(err, this.$t('governance.directReports.errorGettingDirectReportUserInfo'));
      });
    },
    /**
     * Request grants for governance resource
     * @param {Object} params query parameters to pass to request
     */
    async queryResource(params) {
      const response = await getGovernanceGrants(params.grantType, this.directReportUserInfo.userId, params);
      this.resourceItems = response.items;
      this.resourceTotalCount = response.totalCount;
    },
    async revokeResourcesAndCloseModal(requestPayload, resourceId, modalId) {
      try {
        await revokeResourcesFromIGA(requestPayload, resourceId, false);
        this.$bvModal.hide(modalId);
      } catch {
        // no catch statement needed
      }
    },
    tabActivated(tabIndex, prevTabIndex) {
      if (tabIndex > -1) {
        this.$route.params.grantType = this.tabItems[tabIndex].grantType;
        this.tabItems[prevTabIndex].active = false;
        this.tabItems[tabIndex].active = true;
        const { grantType, userId } = this.$route.params;
        window.history.pushState(window.history.state, '', `#/my-reports/${userId}/${grantType}`);
      }
    },
  },
};
</script>
