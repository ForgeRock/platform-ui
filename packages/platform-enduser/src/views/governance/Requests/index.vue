<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BContainer>
    <div class="mt-5">
      <FrHeader
        :title="$t('pageTitles.MyRequests')"
        :subtitle="$t('governance.accessRequests.myRequests.subtitle')" />
      <BCard
        no-body
        class="card-tabs-vertical" />
    </div>
  </BContainer>
</template>

<script>
import {
  BCard,
  BContainer,
} from 'bootstrap-vue';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
/**
 * @description Landing page for Requests
 */
export default {
  name: 'Requests',
  components: {
    BCard,
    BContainer,
    FrHeader,
  },
  data() {
    return {
    };
  },
};
</script>
