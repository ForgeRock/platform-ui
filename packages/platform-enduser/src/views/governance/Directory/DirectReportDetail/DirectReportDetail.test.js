/**
 * Copyright (c) 2023-2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import { setupTestPinia } from '@forgerock/platform-shared/src/utils/testPiniaHelpers';
import * as AccessRequestApi from '@forgerock/platform-shared/src/api/governance/AccessRequestApi';
import * as CommonsApi from '@forgerock/platform-shared/src/api/governance/CommonsApi';
import { mount, flushPromises } from '@vue/test-utils';
import Notifications from '@kyvg/vue3-notification';
import i18n from '@/i18n';
import DirectReportDetail from './index';
import * as DirectoryApi from '@/api/governance/DirectoryApi';

CommonsApi.getIgaAccessRequest = jest.fn().mockImplementation(() => Promise.resolve({
  data: {
    requireRequestJustification: false,
    requireRejectJustification: false,
    requireApproveJustification: false,
    defaultApprover: '',
    allowSelfApproval: false,
  },
}));
AccessRequestApi.saveNewRequest = jest.fn().mockImplementation(() => Promise.resolve({ data: {} }));
CommonsApi.getUserGrants = jest.fn().mockImplementation(() => Promise.resolve({ data: {} }));

describe('DirectReportDetail', () => {
  const setup = (grantType) => {
    setupTestPinia({ user: { userId: 'testId' } });
    return mount(DirectReportDetail, {
      global: {
        plugins: [Notifications, i18n],
        mocks: {
          $route: {
            name: 'DirectReportDetail',
            params: {
              grantType,
              userId: 'reporteeId',
            },
          },
        },
      },
    });
  };

  beforeAll(() => {
    jest.spyOn(DirectoryApi, 'getDirectReportUserInfo').mockReturnValue(Promise.resolve({
      data: {
        givenName: 'Test',
        profileImage: './test/img',
        sn: 'User',
        userName: 'Test.User@forgerock.com',
        _refResourceId: 'id',
      },
    }));
  });

  describe('@renders', () => {
    it('side menu with default tab activated', async () => {
      const wrapper = setup('account');
      await flushPromises();
      const directReportDetailTabApp = findByTestId(wrapper, 'access-tab-account');
      expect(directReportDetailTabApp.attributes().class).toContain('active');
      const directReportDetailTabRole = findByTestId(wrapper, 'access-tab-role');
      expect(directReportDetailTabRole.attributes().class).not.toContain('active');
      const directReportDetailTabEnt = findByTestId(wrapper, 'access-tab-entitlement');
      expect(directReportDetailTabEnt.attributes().class).not.toContain('active');
    });
    it('render correct user info in header', async () => {
      const wrapper = setup('account');
      await flushPromises();
      const directReportDetailHeader = findByTestId(wrapper, 'detail-report-header');
      expect(directReportDetailHeader.find('img').attributes('src')).toBe('./test/img');
      expect(directReportDetailHeader.find('h1').text()).toBe('Test User');
      expect(directReportDetailHeader.find('p').text()).toBe('Test.User@forgerock.com');
    });
    it('right tab on page refresh', async () => {
      const wrapper = setup('entitlement');
      await flushPromises();
      const directReportDetailTabEnt = findByTestId(wrapper, 'access-tab-entitlement');
      expect(directReportDetailTabEnt.attributes().class).toContain('active');
    });
    it('default error message on getDirectReportUserInfo fails', async () => {
      jest.spyOn(DirectoryApi, 'getDirectReportUserInfo').mockReturnValueOnce(Promise.reject());
      const wrapper = setup('account');
      const showErrorSpy = jest.spyOn(wrapper.vm, 'displayNotification');
      await flushPromises();
      expect(showErrorSpy).toHaveBeenCalledTimes(1);
      expect(showErrorSpy).toBeCalledWith('danger', 'Error getting direct report user information.');
    });
  });
});
