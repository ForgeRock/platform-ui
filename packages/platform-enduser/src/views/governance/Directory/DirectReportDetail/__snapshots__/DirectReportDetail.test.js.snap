// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectReportDetail @renders default error message on getDirectReportUserInfo fails: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`DirectReportDetail @renders render correct user info in header: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`DirectReportDetail @renders right tab on page refresh: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;

exports[`DirectReportDetail @renders side menu with default tab activated: console-error-snapshot 1`] = `
[
  [
    "Un-mocked call made to Axios create. All API calls should be mocked",
  ],
  [
    "Un-mocked API call made with Axios. All API calls should be mocked",
  ],
]
`;
